### DISCUSSION TITLE
Is it safe to delete batches in the dead set?

### DISCUSSION THREAD

**Discussion by dhughes** [sidekiq/sidekiq#6011] [Q&A] [answered]
Confession: I've had really bad Sidekiq "hygene" and need advice on how to safely clean things up. 

I have a ton of old batches that have no jobs, so they're undefined. I've recently put in some fixes that should prevent this. I think these undefined batches get automatically cleaned up after 4 weeks. Is that correct? Do I need to do anything else to clean them up?

I've also been looking through the `Sidekiq::Batch::DeadSet`. I have a _lot_ in this set. Like... more than 800,000. I have zero jobs in the dead queue. (Is that the right term?) Some of these batches date back 6 months. Is it safe to just delete them with a script like this?

````
batch_dead_set = Sidekiq::Batch::DeadSet.new.lazy
batch_dead_set
  .each_slice(100) do |set_of_batches|
    set_of_batches.each do |batch|
      batch.delete
    end
end
```

**Comment by mperham** [comment, marked as answer]
0. Batches expire in Redis after 30 days (by default).
1. It's always dangerous to call `batch.delete` yourself because there are no guardrails to ensure data consistency. If a batch is deleted but a batch job still remains in Redis, that job will raise an error if it ever runs again.
2. Iterating through Batch::DeadSet will automatically remove expired batches but will be slow.
3. I'd suggest you call `S::B::DeadSet.new.clear` to remove those 800,000 entries. That size is not intended; I need to think about how to auto-clean that structure.
4. If you remove a batch from the deadset, that's ok but if you delete the batch, that can cause havok per (1).

**Reply by dhughes** [reply]
Thanks, @mperham. This is really helpful information!
